File: lib/src/data/loader.ts
import * as fs from 'fs';
import * as path from 'path';
import * as chokidar from 'chokidar';

import { Budget, BudgetBuilder } from '../core/budget';
import { loadYamlFile, parseYamlContent, YamlData } from './loader_yaml';
import { TransactionStatement, buildTransactionStatementTable } from '../statement/transaction';
import { SummaryStatement, buildSummaryStatementTable } from '../statement/summary';
import { Map3 } from '../utils';

function* readdirSync(file_path: string): Generator<string> {
  if (fs.statSync(file_path).isDirectory()) {
    for (const f of fs.readdirSync(file_path)) {
      const child_file = path.join(file_path, f);
      for (const cf of readdirSync(child_file)) {
        if (cf.match(/\.yaml$/)) {
          yield cf;
        }
      }
    }
  } else {
    yield file_path;
  }
  return;
}

async function realPath(path: string): Promise<string> {
  const stats = await fs.promises.lstat(path);
  return stats.isSymbolicLink() ? await fs.promises.readlink(path) : path;
}

export function listFiles(): string[] {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
  }
  return filePaths;
}

let watchedPath: string | undefined = undefined;
let watcher: chokidar.FSWatcher | undefined = undefined;
let processedBudget: ProcessedBudget | undefined = undefined;

export function unwatchBudgetFiles() {
  watchedPath = undefined;
  watcher?.close();
  watcher = undefined;
  processedBudget = undefined;
}

export interface DataPayload {
  budget: Budget;
  statements: Map<string, Map<string, TransactionStatement>>;
  // owner -> accout name -> month -> summary.
  summaries: Map3<SummaryStatement>;
}

class ProcessedBudget {
  private readonly parsedAccountData = new Map<string, YamlData>();
  budget?: Budget = undefined;
  readonly accoutToMonthToTransactionStatement = new Map<string, Map<string, TransactionStatement>>();
  summaryNameMonthMap = new Map3<SummaryStatement>();

  reProcess() {
    const startTimeMs: number = Date.now();
    const budgetBuilder = new BudgetBuilder();
    for (const [filePath, accountData] of this.parsedAccountData) {
      try {
        loadYamlFile(budgetBuilder, accountData, filePath);
        if (!budgetBuilder.accounts.has(accountData?.name ?? '')) {
          console.warn(`warning: ${filePath} is not an account file.`);
        }
      } catch (e) {
        console.error(`error: Failed to add ${filePath}, ${e}`);
        return;
      }
    }
    this.budget = budgetBuilder.build();
    console.log(`Done building budget for ${this.budget.accounts.size} accounts ` 
        + ` in ${Date.now() - startTimeMs}ms`);

    this.accoutToMonthToTransactionStatement.clear();
    this.summaryNameMonthMap.clear();

    const startBuildTransactionStatementsTimeMs: number = Date.now();
    const transactionStatementTable = buildTransactionStatementTable(this.budget);
    for (const stmt of transactionStatementTable) {
      let monthToStatement = this.accoutToMonthToTransactionStatement.get(stmt.account.name);
      if (!monthToStatement) {
        monthToStatement = new Map<string, TransactionStatement>();
        this.accoutToMonthToTransactionStatement.set(stmt.account.name, monthToStatement);
      }
      monthToStatement.set(stmt.month.toString(), stmt);
    }
    console.log(`Done building ${transactionStatementTable.length} transaction statements ` 
                + ` in ${Date.now() - startBuildTransactionStatementsTimeMs}ms`);

    const startBuildSummaryStatementsTimeMs: number = Date.now();
    this.summaryNameMonthMap = buildSummaryStatementTable(transactionStatementTable);
    let numSummaryStatements = this.summaryNameMonthMap.size;
    console.log(`Done building ${numSummaryStatements} summary statements ` 
                + ` in ${Date.now() - startBuildSummaryStatementsTimeMs}ms`);

    console.log(`Done reprocessing ${this.parsedAccountData.size} file(s) ` 
        + `${transactionStatementTable.length} tran statements and ` 
        + `${numSummaryStatements} summaries in ${Date.now() - startTimeMs}ms`);
  }

  addFolder(rootPath: string) {
    const startTimeMs: number = Date.now();
    for (const filePath of readdirSync(rootPath)) {
      const relativeFilePath = filePath.slice(rootPath.length + 1);
      console.log('Loading ' + relativeFilePath);
      this.addFile(rootPath, relativeFilePath);
    }
    console.log(`Done loading ${this.parsedAccountData.size} file(s) in ${Date.now() - startTimeMs}ms`);
  }

  addFile(rootPath: string, relativeFilePath: string) {
    const content = fs.readFileSync(`${rootPath}/${relativeFilePath}`, 'utf8');
    const accountData = parseYamlContent(content, relativeFilePath);
    if (accountData === undefined) {
      throw Error(
        `Failed to parse ${relativeFilePath} content of size ${content.length} fileStat: ${relativeFilePath}`
      );
    }
    this.parsedAccountData.set(relativeFilePath, accountData);
  }
}

export async function loadBudget(): Promise<DataPayload> {
  if (processedBudget !== undefined) {
    return { 
      budget: processedBudget.budget!, 
      statements: processedBudget.accoutToMonthToTransactionStatement, 
      summaries: processedBudget.summaryNameMonthMap,  
    };
  
  }
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  
  processedBudget = new ProcessedBudget();
  
  const pathToData = await realPath(process.env.TALLY_FILES);
  processedBudget.addFolder(pathToData);
  processedBudget.reProcess();

  if (watcher === undefined || watchedPath !== pathToData) {
    if (watcher) {
      unwatchBudgetFiles();
    }
    watchedPath = pathToData;
    watcher = chokidar
      .watch(pathToData, {
        persistent: true,
        // Multiple change events can be emitted as a larger file is being written.
        awaitWriteFinish: {
          stabilityThreshold: 300,
        },
      })
      .on('all', async (eventType: string, fileStat: fs.Stats | undefined) => {
        if (!fileStat || ['add', 'addDir'].includes(eventType)) {
          return;
        }
        console.log(`${eventType} for ${fileStat}`);
        const startTimeMs: number = Date.now();
        const relativeFilePath = `${fileStat}`.slice(pathToData.length + 1);
        if (processedBudget) {
          processedBudget.addFile(pathToData, relativeFilePath);
          processedBudget.reProcess();
        } else {
          console.error(`Unexpected error: processedBudget is unset`);
        }
        console.log(`info: Rebuilt budget in ${Date.now() - startTimeMs}ms`);
      });
    console.log(`info: Watching ${process.env.TALLY_FILES}`);
  }
  return { 
    budget: processedBudget.budget!, 
    statements: processedBudget.accoutToMonthToTransactionStatement, 
    summaries:  processedBudget.summaryNameMonthMap,  
  };
}

File: server/kt/src/main/kotlin/data/Loader.kt
package com.cherba29.tally.data

import com.cherba29.tally.Map3
import com.cherba29.tally.core.Budget
import com.cherba29.tally.core.BudgetBuilder
import com.cherba29.tally.statement.SummaryStatement
import com.cherba29.tally.statement.TransactionStatement
import com.cherba29.tally.statement.buildSummaryStatementTable
import com.cherba29.tally.statement.buildTransactionStatementTable
import io.github.oshai.kotlinlogging.KotlinLogging
import java.lang.AutoCloseable
import java.nio.file.Path
import java.nio.file.Paths
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlin.io.path.extension
import kotlin.io.path.name
import kotlin.io.path.pathString
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import kotlin.io.path.walk
import kotlin.time.Clock
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

// TODO: probably this is not needed, remove?
fun listFiles(tallyPath: Path): List<String> {
  val filePaths: MutableList<String> = mutableListOf()
  for (filePath in tallyPath.walk()) {
    val relativeFilePath = filePath.relativeTo(tallyPath)
    if (relativeFilePath.extension == "yaml") {
      filePaths.add(relativeFilePath.toString())
    }
  }
  return filePaths.sorted()
}

// TODO: separate this logic into separate module.
class ProcessedBudget {
  private val parsedAccountData = mutableMapOf<String, YamlData>()
  var budget: Budget? = null
  val accountToMonthToTransactionStatement: MutableMap<String, MutableMap<String, TransactionStatement>> =
    mutableMapOf()
  var summaryNameMonthMap = Map3<SummaryStatement>()

  fun reProcess() {
    // TODO: use measure time.
    val startTimeMs: Long = Clock.System.now().toEpochMilliseconds()
    val budgetBuilder = BudgetBuilder()
    val unwantedFiles = mutableListOf<String>()
    for ((filePath, accountData) in parsedAccountData) {
      try {
        loadYamlFile(budgetBuilder, accountData, Paths.get(filePath))
        if (!budgetBuilder.accounts.contains(accountData.name ?: "")) {
          logger.warn { "warning: $filePath is not an account file." }
          unwantedFiles.add(filePath)
        }
      } catch (e: Exception) {
        logger.error { "error: Failed to add $filePath, $e" }
        return
      }
    }
    // Since reprocess is called multiple times, no need reprocess them.
    for (filePath in unwantedFiles) {
      parsedAccountData.remove(filePath)
    }
    budget = budgetBuilder.build()
    logger.info {
      "Done building budget for ${budget?.accounts?.size} accounts  in ${
        Clock.System.now().toEpochMilliseconds() - startTimeMs
      }ms"
    }

    accountToMonthToTransactionStatement.clear()
    summaryNameMonthMap.clear()

    // TODO: use measureTime.
    val startBuildTransactionStatementsTimeMs: Long = Clock.System.now().toEpochMilliseconds()
    val transactionStatementTable = buildTransactionStatementTable(budget!!, owner = null)
    for (stmt in transactionStatementTable) {
      var monthToStatement = accountToMonthToTransactionStatement[stmt.account.name]
      if (monthToStatement == null) {
        monthToStatement = mutableMapOf()
        accountToMonthToTransactionStatement[stmt.account.name] = monthToStatement
      }
      monthToStatement[stmt.month.toString()] = stmt
    }
    logger.info {
      "Done building ${transactionStatementTable.size} transaction statements in ${
        Clock.System.now().toEpochMilliseconds() - startBuildTransactionStatementsTimeMs
      }ms"
    }

    val startBuildSummaryStatementsTimeMs: Long = Clock.System.now().toEpochMilliseconds()
    summaryNameMonthMap = buildSummaryStatementTable(transactionStatementTable, selectedOwner = null)
    val numSummaryStatements = summaryNameMonthMap.size
    logger.info {
      "Done building $numSummaryStatements summary statements in ${
        Clock.System.now().toEpochMilliseconds() - startBuildSummaryStatementsTimeMs
      }ms"
    }

    logger.info {
      "Done reprocessing ${parsedAccountData.size} file(s) ${transactionStatementTable.size} tran statements and $numSummaryStatements summaries in ${
        Clock.System.now().toEpochMilliseconds() - startTimeMs
      }ms"
    }
  }

  fun addFile(rootPath: Path, relativeFilePath: Path) {
    val content = rootPath.resolve(relativeFilePath).readText(Charsets.UTF_8)
    val accountData = parseYamlContent(content, relativeFilePath)
    if (accountData == null) {
      throw IllegalStateException(
        "Failed to parse $relativeFilePath content of size ${content.length} fileStat: $relativeFilePath"
      )
    }
    parsedAccountData[relativeFilePath.toString()] = accountData
  }

  companion object {
    private val logger = KotlinLogging.logger {}
  }
}

data class DataPayload(
  val budget: Budget,
  val statements: Map<String, Map<String, TransactionStatement>>,
  // owner -> account name -> month -> summary.
  val summaries: Map3<SummaryStatement>,
)

class Loader(tallyFilesPath: Path): AutoCloseable {
  private val watchedPath: Path = tallyFilesPath.toRealPath()

  private val processedBudget = ProcessedBudget()

  private val dataLock = ReentrantLock()
  private val signalLock = ReentrantLock()
  private val condition = signalLock.newCondition()
  private var processedOn: Long = 0
  val loadedOn: Long get() = processedOn


  private val watcherJob: Job = watchedPath.watchedEventFlow {
    it.extension == "yaml" && !ignorePathRegex.containsMatchIn(it.pathString)
  }.onEach { (relativePath: Path?, reprocess: Boolean) ->
    val startTimeMs: Long = Clock.System.now().toEpochMilliseconds()
    if (relativePath != null) {
      processedBudget.addFile(watchedPath, relativePath)
    }
    if (reprocess) {
      dataLock.withLock {
        processedBudget.reProcess()
        if (processedOn == 0L) {
          signalLock.withLock {
            condition.signalAll()
          }
        }
        // TODO: use provided clock instead.
        processedOn = Clock.System.now().toEpochMilliseconds()
      }
      logger.info { "info: Rebuilt budget in ${processedOn - startTimeMs}ms" }
    }
  }.launchIn(CoroutineScope(Dispatchers.IO + SupervisorJob()))

  init {
    logger.info { "Watching $watchedPath" }
  }

  val budget: DataPayload get() {
    if (processedOn == 0L) {
      // If first time, make sure the data has been loaded before proceeding.
      signalLock.withLock {
        while (processedOn == 0L) {
          condition.await()
        }
      }
    }
    dataLock.withLock {
      return DataPayload(
        budget = processedBudget.budget!!,
        statements = processedBudget.accountToMonthToTransactionStatement,
        summaries = processedBudget.summaryNameMonthMap,
      )
    }
  }

  override fun close() {
    watcherJob.cancel()
  }

  companion object {
    val ignorePathRegex = Regex("(^_)|(/_)")
    private val logger = KotlinLogging.logger {}
  }
}
